solr/core/src/test/org/apache/solr/search/TestMainQueryCaching.java

```diff
@@ -0,0 +1,313 @@
+    final boolean consultMatchAllDocs;
+    final boolean insertFilterCache;
+    final boolean skipSort;
+    if (MATCH_ALL_DOCS_QUERY.equals(q)) {
+      consultMatchAllDocs = true;
+      insertFilterCache = false;
+      skipSort = true;
+    } else {
+      consultMatchAllDocs = false;
+      insertFilterCache = USE_FILTER_FOR_SORTED_QUERY;
+      skipSort = USE_FILTER_FOR_SORTED_QUERY;
+    }
+    assertMetricCounts(response, consultMatchAllDocs, insertFilterCache ? 1 : 0, skipSort ? 0 : 1, skipSort ? 1 : 0);
+  }
+
+  private static void assertMetricCounts(String response, boolean matchAllDocs, int expectFilterCacheInsertCount, int expectFullSortCount, int expectSkipSortCount) {
+    assertMetricCounts(response, matchAllDocs, expectFilterCacheInsertCount, expectFullSortCount, expectSkipSortCount, matchAllDocs ? ALL_DOCS : MOST_DOCS);
+  }
+
+  private static void assertMetricCounts(String response, boolean matchAllDocs, int expectFilterCacheInsertCount,
+                                         int expectFullSortCount, int expectSkipSortCount, int expectNumFound) {
+    Map<?, ?> res = (Map<?, ?>) fromJSONString(response);
+    Map<?, ?> body = (Map<?, ?>) (res.get("response"));
+    SolrCore core = h.getCore();
+    assertEquals("Bad matchAllDocs insert count", (matchAllDocs ? 1 : 0), coreToMatchAllDocsInsertCount(core));
+    assertEquals("Bad filterCache insert count", expectFilterCacheInsertCount, coreToInserts(core));
+    assertEquals("Bad full sort count", expectFullSortCount, coreToSortCount(core, "full"));
+    assertEquals("Bad skip sort count", expectSkipSortCount, coreToSortCount(core, "skip"));
+    assertEquals("Should have exactly " + expectNumFound, expectNumFound, (long) (body.get("numFound"))); // sanity check
+  }
+
+  @Test
+  public void testConcurrentMatchAllDocsInitialization() throws Exception {
+    final int nThreads = 20;
+    final ExecutorService executor = ExecutorUtil.newMDCAwareFixedThreadPool(nThreads, new SolrNamedThreadFactory(getTestName()));
+    final Future<?>[] followup = new Future<?>[nThreads];
+    for (int i = 0; i < nThreads; i++) {
+      final int myI = i;
+      followup[i] = executor.submit(() -> {
+        try {
+          // NOTE: we use cursorMark=* here because it prevents consulting the queryResultCache, which can interfere
+          // with DocSet fetching (which is what we care about in this test).
+          String response = JQ(req("q", MATCH_ALL_DOCS_QUERY, "request_id", Integer.toString(myI), "cursorMark", "*", "sort", "id asc"));
+          Map<?, ?> res = (Map<?, ?>) fromJSONString(response);
+          Map<?, ?> body = (Map<?, ?>) (res.get("response"));
+          assertEquals("Should have exactly " + ALL_DOCS, ALL_DOCS, (long) (body.get("numFound"))); // sanity check
+        } catch (Exception ex) {
+          throw new RuntimeException(ex);
+        }
+      });
+    }
+    try {
+      for (Future<?> f : followup) {
+        f.get(); // to access exceptions/errors
```

Comment:
`assertNull` here

Reply:
I'm curious why? If the issue is _swallowing_ errors/exceptions, these would not be returned by `.get()` anyway, but will instead be thrown wrapped in `ExecutionException`, iiuc.